### PR TITLE
replace libzmq with websockets (holochain-rust side)

### PR DESCRIPTION
**Pull request by neonphog**
- [x] I have added a summary of my changes to the changelog

see companion n3h pr:
https://github.com/holochain/n3h/pull/51

also see companion docs pr:
https://github.com/holochain/docs-pages/pull/19

Still need to do some unit testing. Will work on that in a separate branch while review proceeds.

### CODE REVIEW

File: net/src/ipc_net_worker.rs [hunk 1]
```diff
@@ -327,6 +322,7 @@ impl IpcNetWorker {
             }"#;
 }
 
+/*
```
Comment (thedavidmeister) [inline net/src/ipc_net_worker.rs]: why are we doing this?

if the tests aren't relevant at all then let's delete them

if they are relevant but need fixing, use the "broken-tests" compiler attribute to elide them

File: net/src/p2p_network.rs [hunk 2]
```diff
@@ -87,6 +87,7 @@ impl NetSend for P2pNetwork {
 mod tests {
     use super::*;
 
+    /*
```
Comment (thedavidmeister) [inline net/src/p2p_network.rs]: delete?

**Comment by thedavidmeister** [comment]
i've put up a PR against the docs repo ready for when this lands https://github.com/holochain/docs-pages/pull/19

**Comment by sphinxc0re** [comment]
Did you remove the vendored zmq libs?

File: cli/src/cli/test.rs [hunk 3]
```diff
@@ -87,6 +87,7 @@ pub fn test(
 }
 
 #[cfg(test)]
+#[cfg(feature = "broken-tests")]
```
Comment (lucksus) [inline cli/src/cli/test.rs]: Why?

File: net/src/ipc/transport_wss/mod.rs [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+    fn close(&mut self, id: TransportId) -> TransportResult<()> {
+        if let Some(mut info) = self.stream_sockets.remove(&id) {
+            self.priv_close_one(&mut info)?;
+        }
+        Ok(())
+    }
+
+    /// close all currently tracked connections
+    fn close_all(&mut self) -> TransportResult<()> {
+        let mut err: TransportResult<()> = Ok(());
```
[the remark this answers is not in the document]
Comment (neonphog) [reply net/src/ipc/transport_wss/mod.rs]: lol... made sense when I wrote it, but yeah, that's totally confusing

File: net/src/ipc_net_worker.rs (hunk 1 above)
Comment (neonphog) [reply net/src/ipc_net_worker.rs]: went ahead and deleted, it'll be easier to rewrite from scratch when I take that pass.

File: net/src/ipc/transport_wss/mod.rs [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+        while !self.stream_sockets.is_empty() {
+            let key = self
+                .stream_sockets
+                .keys()
+                .next()
+                .expect("should not be None")
+                .to_string();
+            if let Some(mut info) = self.stream_sockets.remove(&key) {
+                if let Err(e) = self.priv_close_one(&mut info) {
+                    err = Err(e.into());
```
[the remark this answers is not in the document]
Comment (neonphog) [reply net/src/ipc/transport_wss/mod.rs]: I indeed had that thought when I implemented this. Switched to returning a vec of errors: :+1:

File: net/src/ipc/transport_wss/mod.rs [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+    }
+
+    /// this should be called frequently on the event loop
+    /// looks for incoming messages or processes ping/pong/close events etc
+    fn poll(&mut self) -> TransportResult<(DidWork, Vec<TransportEvent>)> {
+        let mut did_work = false;
+
+        if self.priv_process_stream_sockets()? {
+            did_work = true
+        }
```
[the remark this answers is not in the document]
Comment (neonphog) [reply net/src/ipc/transport_wss/mod.rs]: hehe, there were originally more steps to this function... updated : )

File: net/src/ipc/transport_wss/mod.rs [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+        Ok((did_work, self.event_queue.drain(..).collect()))
+    }
+
+    /// send a message to one or more remote connected nodes
+    fn send(&mut self, id_list: &[&TransportIdRef], payload: &[u8]) -> TransportResult<()> {
+        for id in id_list {
+            if let Some(info) = self.stream_sockets.get_mut(&id.to_string()) {
+                info.send_queue.push(payload.to_vec());
+            } else {
+                return Err(TransportError(format!("bad id: {}", id)));
```
[the remark this answers is not in the document]
Comment (neonphog) [reply net/src/ipc/transport_wss/mod.rs]: actually... we're looking up that key from the very hash_map we're fetching from, it shouldn't be possible to get into that error state, i just removed it.

File: net/src/ipc_net_worker.rs [hunk 8]
```diff
@@ -140,124 +105,153 @@ impl IpcNetWorker {
...
+fn wait_connect(
+    socket: &mut TransportWss<std::net::TcpStream>,
+    uri: &str,
+) -> NetResult<Vec<TransportEvent>> {
+    let mut out = Vec::new();
+
+    socket.connect(&uri)?;
+
+    let start = std::time::Instant::now();
+    while start.elapsed().as_millis() < 5000 {
```
[the remark this answers is not in the document]
Comment (neonphog) [reply net/src/ipc_net_worker.rs]: actually these are slightly unrelated. The heartbeat timeout is in play once the connection is establish, the connection itself is unbound within the transport code. But i like the idea of the constant, went ahead and implemented... we can always change the values separately if we have need to.

**Comment by neonphog** [comment]
@thedavidmeister - yep, removed a couple additional tests - tackling them from scratch will be cleaner anyway.

Let me know if this looks good : )

**Pull request merged**